Vwitter/SpotifyAPI/SpotifyAPIManager.m

```diff
@@ -0,0 +1,212 @@
+        if (!error) {
+            NSLog(@"track playing");
+        }
+        else {
+            NSLog(@"cell track failed to play %@", error.localizedDescription);
+        }
+    }];
+}
+
+- (void)pause {
+    [self.appRemote.playerAPI pause:^(id  _Nullable result, NSError * _Nullable error) {
+        if (!error) {
+            NSLog(@"track paused");
+
+        }
+        else {
+            NSLog(@"track not paused, %@", error.localizedDescription);
+        }
+    }];
+    
+}
+
+- (void)getTracks:(NSString *)searchString withCompletion:(SpotifyTrackCompletion)completion {
+    if (!self.hasValidSpotifyAuthorization) {
+        [self authorizeSpotify];
+    }
+    
+    NSString *bearerToken = [[NSString alloc] initWithFormat:@"Bearer %@", self.appRemote.connectionParameters.accessToken];
+    
+    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
+    
+    NSString *convertedSearchString = [searchString stringByReplacingOccurrencesOfString:@" " withString:@"%20"];
+    
+    NSString *urlString = [NSString stringWithFormat:@"%@/%@/%@", @"https://api.spotify.com/v1/search?q=", convertedSearchString, @"&type=track"];
+    NSURL *url = [NSURL URLWithString:urlString];
+
+    [request setValue:bearerToken forHTTPHeaderField:@"Authorization"];
+    [request setURL:url];
+
+    NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
+      [[session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+        if (!error) {
+            NSMutableArray *spotifyTracksArray = [[NSMutableArray alloc] init];
+            
+            NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
+            // JSON with song is here
+            NSLog(@"JSON: %@", json);
+            NSArray *tracksArray = json[@"tracks"][@"items"];
+            NSArray *castedTracksArray = CAST_TO_CLASS_OR_NIL(tracksArray, NSArray);
+            
+            for (id track in castedTracksArray) {
+                SpotifyTrack *currentTrack = [[SpotifyTrack alloc] initWithDictionary:track];
+                [spotifyTracksArray addObject:currentTrack];
+            }
+            
+            completion(spotifyTracksArray, nil);
+            
+        }
+        else {
+            NSLog(@"searching tracks had an error %@", error.localizedDescription);
+            completion(nil, error);
+        }
+    }] resume];
+
+}
+
+- (void)getDurationWithCompletion:(DurationCompletion)completion {
+    
+    [self.appRemote.playerAPI getPlayerState:^(id  _Nullable result, NSError * _Nullable error) {
+        if (!error) {
+//            SPTAppRemotePlayerState *playerState = CAST_TO_CLASS_OR_NIL(result, SPTAppRemotePlayerState);
+            id<SPTAppRemotePlayerState> playerState = result;
+            completion(playerState.track.duration, nil);
```

Comment:
This code is a bit dangerous. `duration` is a primitive. This means that it can literally never be `nil` like an object/pointer can. Instead, if spotify returns `nil` playerState, when you call `playerState.track.duration`, ObjC will give us `0`. 

You should use something more like this pattern

```
if (!error) {
  id<SPTAppRemotePlayerState> playerState = CAST_TO_PROTOCOL_OR_NIL(result, @protocol(SPTAppRemotePlayerState));
  if (!playerState) {
    completion(0, [NSError errorWith.... // (create a new error object)
  } else {
    completion(playerState.track.duration, nil);
  }
} else {
  // log it and give completion the error object
}
```